### PR TITLE
Scripts/Zul'Gurub 22974 high priestess marli hatching eggs

### LINKED ISSUE #22874: >> Applying update "2018_12_09_00_hotfixes.sql" '3BC8C33'... ERROR 1054 (42S22) at line 86: Unknown column 'PosX' in 'artifact_power'
**Description:**

cant import hotfix database updates and without them the server wont run. i created the mysql from the source code didnt import anything just opened bnetserver and world server and let them auto populate the empty databases but it always fails at 

>> Applying update "2018_12_09_00_hotfixes.sql" '3BC8C33'...
ERROR 1054 (42S22) at line 86: Unknown column 'PosX' in 'artifact_power'

this is extremely frustrating ive even edited that update to remove the posx in artifact power and ran it threw Heidisql with no success 

*Current behaviour:**

>> Applying update "2018_12_09_00_hotfixes.sql" '3BC8C33'...
ERROR 1054 (42S22) at line 86: Unknown column 'PosX' in 'artifact_power'

ive tried many differnt ways and they all end the same this update stops me yet i need them or the server returns 

In mysql_stmt_prepare() id: 16, sql: "UPDATE creature SET MovementType = ? WHERE guid = ?"
Unknown column 'guid' in 'where clause'
In mysql_stmt_prepare() id: 19, sql: "UPDATE creature SET position_x = ?, position_y = ?, position_z = ?, orientation = ? WHERE guid = ?"
Unknown column 'guid' in 'where clause'
In mysql_stmt_prepare() id: 20, sql: "UPDATE creature SET spawndist = ?, MovementType = ? WHERE guid = ?"
Unknown column 'guid' in 'where clause'
In mysql_stmt_prepare() id: 21, sql: "UPDATE creature SET spawntimesecs = ? WHERE guid = ?"
Unknown column 'guid' in 'where clause'
In mysql_stmt_prepare() id: 53, sql: "DELETE FROM creature WHERE guid = ?"
Unknown column 'guid' in 'where clause'
In mysql_stmt_prepare() id: 62, sql: "INSERT INTO creature (guid, id , map, spawnMask, phaseMask, modelid, equipment_id, position_x, position_y, position_z, orientation, spawntimesecs, spawndist, currentwaypoint, curhealth, curmana, MovementType, npcflag, unit_flags, dynamicflags) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
Unknown column 'guid' in 'field list'
In mysql_stmt_prepare() id: 65, sql: "INSERT INTO gameobject (guid, id, map, spawnMask, phaseMask, position_x, position_y, position_z, orientation, rotation0, rotation1, rotation2, rotation3, spawntimesecs, animprogress, state) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
Unknown column 'phaseMask' in 'field list'

this is extremely frustrating ive even edited that update to remove the posx in artifact power and ran it threw Heidisql with no success 

**Expected behaviour:**

the hotfixes sql update 12-9-00 needs to be fixed 

**Steps to reproduce the problem:**

1. Step 1 Download and compile the master source code 
2. Step 2 Copy all into one folder with the dbc,maps,mmaps,vmaps,gt,worldserver.exe/conf and bnetserver.exe/conf along with the lastest 7.3.5 world and hotfixes database sqls 
3. Step 3 Open create_mysql.sql in sql/create inside the source codes directory 
4. Step 4 Run Bnetserver.exe and Worldserver.exe allow it to populate and update the databases until it envitably crashes at Error 1054  12_9_00_hotfixes.sql  has unknown entity 'PosX' in colomn 'artifact_power'

**Branch(es):**

master

**TC rev. hash/commit:** 

CHANGEME Copy the result of server debug command (if you need to run it from client get prat addon)

**Operating system:** windows 10

### LINKED ISSUE #22974: Creature/script: High Priestess Mar'li
**Description:** This boss https://www.wowhead.com/npc=14510/high-priestess-marli must hatch spider eggs https://www.wowhead.com/object=179985/spider-egg that summon a spider.

In the current script the spider is summoned manually:
if (Creature* spider = me->SummonCreature(NPC_SPIDER, target->GetPositionX(), target->GetPositionY(), target->GetPositionZ(), 0, TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT, 15000))

Spell that must be used to hatch eggs: https://woehead.way-of-elendil.fr/?spell=24082

Hatched eggs must respawn after a wipe.

**Branch(es):** 3.3.5

**TC rev. hash/commit:** rev. 813cf761fcf8

### PR DESCRIPTION
**Pull request by kuskmen**
**Changes proposed:**

- Boss should use spider eggs in order to spawn it's minion spiders.  
- Spider hatch event should occur only during the human phase.
- Summoned spiders and hatched eggs should respawn after event encounter resets.
- Summoned spiders should have better targeting than they used to.

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #22974

**Tests performed:** It builds and I've tested it in-game.

**Known issues and TODO list:** None.

### CODE REVIEW

File: src/server/game/AI/ScriptedAI/ScriptedCreature.cpp [hunk 1]
```diff
@@ -123,6 +123,17 @@ void SummonList::DoActionImpl(int32 action, StorageType const& summons)
     }
 }
 
+std::list<ObjectGuid> SummonList::GetGUIDs() const
```
Comment (kuskmen) [inline src/server/game/AI/ScriptedAI/ScriptedCreature.cpp]: Is there any way making this return some sort of generic container like IEnumerable in C#?

File: src/server/game/Entities/Object/Object.h [hunk 2]
```diff
@@ -396,7 +397,10 @@ class TC_GAME_API WorldObject : public Object, public WorldLocation
         void SummonCreatureGroup(uint8 group, std::list<TempSummon*>* list = nullptr);
 
         Creature*   FindNearestCreature(uint32 entry, float range, bool alive = true) const;
+        std::list<Creature*> FindAllCreaturesByEntry(uint32 entry, float range, bool alive = true) const;
```
Comment (kuskmen) [inline src/server/game/Entities/Object/Object.h]: Is there any way making this return some sort of generic container like IEnumerable in C#?

File: src/server/game/Grids/Notifiers/GridNotifiers.h [hunk 3]
```diff
@@ -757,6 +757,29 @@ namespace Trinity
...
     };
 
+    class NearestReadyStateGameObjectEntryInObjectRangeCheck
+    {
+        public:
+            NearestReadyStateGameObjectEntryInObjectRangeCheck(WorldObject const& obj, uint32 entry, float range) : i_obj(obj), i_entry(entry), i_range(range) { }
+
+            bool operator()(GameObject* go)
+            {
+                if (go->GetEntry() == i_entry && i_obj.IsWithinDistInMap(go, i_range) && go->getLootState() == GO_READY)
```
Comment (kuskmen) [inline src/server/game/Grids/Notifiers/GridNotifiers.h]: I've tried inheriting from more common check only adding concrete check here like so

```
if(Base::operator()(go) && concrete checks..)
```

Didn't work the way I thought it will, should I try it again or this also works?

**Comment by kuskmen** [comment]
circleci: 
> /home/circleci/project/src/server/game/Grids/Notifiers/GridNotifiers.h:1326:59: fatal error: field 'i_range' will be initialized after field 'i_creatureGuids' [-Wreorder]
            : i_obj(obj), i_entry(entry), i_alive(alive), i_range(range), i_creatureGuids(creatureGuids) { }

Dunno whats the problem with that if someone elaborates I will fix.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp [hunk 4]
```diff
@@ -77,142 +81,187 @@ enum Misc
...
-            }
+                    summons.DespawnAll();
+                    //_hatchedEggsGuids.clear();
+                    _Reset();
+                }
+                
+                //void SpellHitTarget(GameObject* target, SpellInfo const* /*spellInfo*/) override
+                //{
+                //    if(target->GetEntry() == GOB_SPIDER_EGG)
+                //        _hatchedEggsGuids.insert(target->GetGUID());
```
Comment (illfated) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: If you intend to stop using part of the code, TC recommends that you remove it instead of commenting it out.

File: src/server/game/Grids/Notifiers/GridNotifiers.h [hunk 5]
```diff
@@ -1296,6 +1319,33 @@ namespace Trinity
             NearestCreatureEntryWithLiveStateInObjectRangeCheck(NearestCreatureEntryWithLiveStateInObjectRangeCheck const&) = delete;
     };
 
+    class NearestCreatureEntryNotInListCheck
+    {
+    public:
+        NearestCreatureEntryNotInListCheck(WorldObject const& obj, uint32 entry, bool alive, float range, std::list<ObjectGuid> creatureGuids)
+            : i_obj(obj), i_entry(entry), i_alive(alive), i_range(range), i_creatureGuids(creatureGuids) { }
```
Comment (illfated) [inline src/server/game/Grids/Notifiers/GridNotifiers.h]:
```cpp
/home/circleci/project/src/server/game/Grids/Notifiers/GridNotifiers.h:1326:59:
 fatal error: field 'i_range' will be initialized after field 'i_creatureGuids' [-Wreorder]
            : i_obj(obj), i_entry(entry), i_alive(alive), i_range(range), i_creatureGuids(creatureGuids) { }
                                                          ^
```
You should rearrange the last 2 arguments: `, i_creatureGuids(creatureGuids, i_range(range)) { }`

**Comment by illfated** [comment]
Cosmetic — suggestion for your PR title:
**Scripts/Zul'Gurub: High Priestess Mar'li hatching spider eggs**

Do you really need to change so much in the Core files (`/src/server/game/`) to make the script work?
(I don't know, I am not a real coder.)

BTW, you need at least to correct your issue link to #22974 from #22874 (completely different & closed issue)

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp (hunk 4 above)
Comment (Faq) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: prolly PR is draft

Comment (illfated) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: Would be nice if the Github Pull Request Draft feature had been used to make it clear.

File: src/server/game/AI/ScriptedAI/ScriptedCreature.cpp (hunk 1 above)
Comment (ccrs) [reply src/server/game/AI/ScriptedAI/ScriptedCreature.cpp]: What for?

Also bear in mind what the impl of this method is actually doing: copying the values and referencing em as copy.
Also also, better use vectors, GuidVector is the typename IIRC

File: src/server/game/Grids/Notifiers/GridNotifiers.h (hunk 3 above)
Comment (ccrs) [reply src/server/game/Grids/Notifiers/GridNotifiers.h]: whats the purpose you were aiming for?

**Comment by ccrs** [comment]
https://github.com/TrinityCore/TrinityCore/pull/23212#issuecomment-487295620

variable initialization order error: you must initialize in the same order you declare em

**Comment by Warpten** [comment]
To clarify:  the order of initializers execution is based on the order in which members are declared. Rules are specified by the standard, summary here https://en.cppreference.com/w/cpp/language/initializer_list

File: src/server/game/Entities/Object/Object.h (hunk 2 above)
Comment (Warpten) [inline src/server/game/Entities/Object/Object.h]: Not sure what the point would be. Lazy initialization? In 99% of the situations needed for scripts there is a solution that doesn't require grid lookups

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp (hunk 4 above)
Comment (kuskmen) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: How can I use this feature @illfated, kinda new to the github organizational features

File: src/server/game/Entities/Object/Object.h (hunk 2 above)
Comment (kuskmen) [reply src/server/game/Entities/Object/Object.h]: @Warpten , so the idea was that this boss spawns multiple creatures of one entry using a spell/gob so in order targeting to work properly after spawned I had to have a way of getting all unique creatures of entry and then script them accordingly. I have to admit this is my first script so additional feedback/guidance is desired, also there might be existing way of achieving what I did that I missed, but in general this was the idea.

File: src/server/game/AI/ScriptedAI/ScriptedCreature.cpp (hunk 1 above)
Comment (kuskmen) [reply src/server/game/AI/ScriptedAI/ScriptedCreature.cpp]: I am not sure how you are guys doing, isn't methods supposed to be at most generic as possible? 
Yes, I will change it to vector.

File: src/server/game/Grids/Notifiers/GridNotifiers.h (hunk 3 above)
Comment (kuskmen) [reply src/server/game/Grids/Notifiers/GridNotifiers.h]: @ccrs , well I thought that inheriting will remove a lot of boilerplate code and reduce duplications as well

**Comment by kuskmen** [comment]
I can't tell if I NEED to change that many core files in order my script to work, I can only tell that this way it works better than it did before imo, but since this is my first scripting I expect other experienced developers to have other to say.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp (hunk 4 above)
Comment (illfated) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: Sorry for being unclear about "the Github Pull Request Draft feature". That feature is only available when opening a pull request, as a drop-down menu, on the button, where you can choose to publish the PR as a Draft instead of being ready for review (but you can convert the PR to "Ready for review" later).

The main intention here was to point out that it is normally recommended to remove the lines of code instead of commenting it out, at least if there is no important point in keeping the code commented out (if the code needs to be used later).

Comment (kuskmen) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: I see, will use it in my future requests, regarding commented code. Problem is that I am unexperienced in scripting yet and how it really should look like is that the boss casts the spell, hatches the egg and after reset event eggs are respawned and all the commented code works on 90% except the respawn part, I tried figuring it out for like 3-4 days debugging but I couldn't so I left it here in case someone knows what might be the reason behind all of this and perhaps point me to right direction, otherwise if maintainers accept this solution as it is I will delete all the unnecessary code before merging :)

File: src/server/game/Entities/Object/Object.h (hunk 2 above)
Comment (Warpten) [reply src/server/game/Entities/Object/Object.h]: i think grid lookups arent as bad as they used to be but generally speaking all you need is to hook JustSummoned in the summoner and store a summon list

f the boss isn't the one summoning the unit (either directly or via a spell), script the summoned unit with a IsSummonedBy hook, retrieve a pointer to the boss from instance script, and call BossAI::JustSummoned(me).

Example:

https://github.com/TrinityCore/TrinityCore/blob/master/src/server/scripts/Northrend/ChamberOfAspects/RubySanctum/boss_halion.cpp#L588

https://github.com/TrinityCore/TrinityCore/blob/master/src/server/scripts/Northrend/ChamberOfAspects/RubySanctum/boss_halion.cpp#L1022

Comment (kuskmen) [reply src/server/game/Entities/Object/Object.h]: @Warpten, thanks for the suggestions, I remember trying JustSummoned hook but as you said its not the boss who summons the spiders its the effect of game object, anyway, I think I am already keeping a summon list and despawning afterwards anyway, what is this all about? :) This method I am using in order to get all unique creatures nearby by given entry and order them to start attack one by one.

**Comment by Aokromes** [comment]
i suggest you to split this PR on 2, the core files and the fix per se.

**Comment by kuskmen** [comment]
@Aokromes okay, will do it whenever I can, should I close this one or you will first merge core files PR then I merge 3.3.5 into this one here?

File: src/server/game/Entities/Object/Object.h (hunk 2 above)
Comment (jackpoz) [reply src/server/game/Entities/Object/Object.h]: > script the summoned unit with a IsSummonedBy hook, retrieve a pointer to the boss from instance script, and call BossAI::JustSummoned(me).

what happened to this ? this will not require any change to "game" project

**Comment by jackpoz** [comment]
imo you should try to achieve what you need without adding new code to the "game" project as there are other ways (like the hook Warpten suggested)

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp [hunk 6]
```diff
@@ -77,142 +81,187 @@ enum Misc
...
+                //}
 
-            void JustEngagedWith(Unit* /*who*/) override
-            {
-                _JustEngagedWith();
-                events.ScheduleEvent(EVENT_SPAWN_START_SPIDERS, 1s, 0, PHASE_ONE);
-                Talk(SAY_AGGRO);
-            }
+                void JustDied(Unit* /*killer*/) override
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: are these just formatting changes ? it makes it hard to review the PR, is it possible to revert the formatting changes ?

Comment (kuskmen) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: Yea I totally agree, must have clicked Ctrl + K + D accidentally leading into auto formatting, will fix this.

File: src/server/game/Entities/Object/Object.h (hunk 2 above)
Comment (kuskmen) [reply src/server/game/Entities/Object/Object.h]: Well, it turns out that Spider Egg is summoning the spiders (it all starts from [here](https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Spells/SpellEffects.cpp#L2231) which leads to [here](https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Object/Object.cpp#L1972) where `ToUnit()` call fails because Spider Egg is gameobject not a unit ultimately failing to set summoner for this tempsummon (the spider) then this check over [here](https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Creature/TemporarySummon.cpp#L207) fails and does not raise the aforementioned hooks.

@Warpten , @jackpoz 

Question is: Is this adequate behavior of the core or we are missing functionality? (To me it sounds that it is possible for gameobject to be summoner but I am new in the project)

Comment (Warpten) [reply src/server/game/Entities/Object/Object.h]: You retrieve a pointer to the boss within IsSummonedBy and call BossAI::JustSummoned, see the second link i pasted. Almost everything on Halion is not summoned by Halion

Comment (Killyana) [reply src/server/game/Entities/Object/Object.h]: Yes game object could be a summoner, but there's an issue, https://github.com/TrinityCore/TrinityCore/issues/23131

Comment (kuskmen) [reply src/server/game/Entities/Object/Object.h]: Yes, I assume we need logic in the core to handle gameobject summoners so they can properly raise further hooks afterwards.

Comment (kuskmen) [reply src/server/game/Entities/Object/Object.h]: Its not that simple, its not the problem of `IsSummonedBy` it is the problem of gameobject not being able to be "owner" of the summoned unit it has from the spell it (the gameobject) casts, which clearly the core does not support and as I see it requires a lot of work.

Comment (jackpoz) [reply src/server/game/Entities/Object/Object.h]: better fix the gameobject summoning in another PR first then

Comment (kuskmen) [reply src/server/game/Entities/Object/Object.h]: @jackpoz I think it will require a lot of core changes and I don't think I will be the best person to do it since I am new to this project but I will try to open one issue explaining whats expected and what is missing imo.

Comment (kuskmen) [reply src/server/game/Entities/Object/Object.h]: here it is https://github.com/TrinityCore/TrinityCore/issues/23242

**Comment by kuskmen** [comment]
@Aokromes , @jackpoz  I don't know what's the state and what should I do more, can you elaborate so I can help push this pull request to source. 

Is there anything I can be of help?

**Comment by illfated** [comment]
Upon browsing through the PR files now, it looks more manageable than the impression I got 10 days ago, and that it can be reviewed as it is now. Then again, the TC owners and team members will decide how to proceed with your PR. Looks quite OK so far, though. As long as you can handle the reviews, it is all good, I suppose.

**Review by ghost — APPROVED** [review]
The other reviews may still need some follow-up work.

**Comment by jackpoz** [comment]
@kuskmen please put this PR on hold until https://github.com/TrinityCore/TrinityCore/issues/23242 is fixed . If you have time you could try implementing that feature in a separate PR and see if anyone has time to help you with that.
I put it on top of my issue list so I might take a look at it this weekend.

**Comment by kuskmen** [comment]
@jackpoz hehe, okay its no problem, I am just trying to take ownership for my babies (pull requests :D) and secure their merging safely :)

**Comment by xvwyh** [comment]
My god does this PR contains A LOT of unnecessary code.

Causing 4 eggs to hatch is achieved solely via spell 24083 "Hatch Eggs"
```
MaxTargets: 4
SpellRange (Id 13) "Anywhere": 0 - 50000
Id: 86 (SPELL_EFFECT_ACTIVATE_OBJECT)
Targets (18, 52) (TARGET_DEST_CASTER, TARGET_GAMEOBJECT_DEST_AREA)
EffectMiscValueA = 8 (Open)
```
Just add a script to it and sort the targets by distance to caster in `OnObjectAreaTargetSelect` hook.

And causing 1 egg to hatch is done via spell 24082 "Hatch Spider Egg"
```
SpellRange (Id 13) "Anywhere": 0 - 50000
Id: 86 (SPELL_EFFECT_ACTIVATE_OBJECT)
BasePoints = 1 DieSides = 1
Targets (40, 0) (TARGET_GAMEOBJECT_NEARBY_ENTRY, NO_TARGET)
EffectMiscValueA = 8 (Open)
```
Said eggs - GO 179985 - have "consumable" property set, meaning they will despawn when they are "opened" by the aforementioned spells, and subsequent casts of the spells shouldn't target despawned GOs for obvious reason.

Just don't forget to add targeting conditions for both spells to allow them to only hit GO 179985.

Why are you adding `FindAllCreaturesByEntry` when you could just add a script to hatched spiders (oh hey, you already have it, it's called `npc_spawn_of_marli`) and aggro them at players from its `IsSummonedBy` method. Even if you don't want to do that, why are you adding `FindAllCreaturesByEntry` anyway when you have `GetCreatureListWithEntryInGrid`?

TL;DR: None of the core changes in this PR are needed.

**Comment by kuskmen** [comment]
@xvwyh , all of your questions have one simple answer. - Its my first pull request.

Thanks for the review, I will try to update my PR appropriately, I was also not so happy having so much changes in core.

**Comment by xvwyh** [comment]
> If you've read all convo in the PR you will know why.

Oh, I see. Hehe, another consequence of GO-casting. Yeah, I agree, you better wait until #23242 is fixed (that's gonna be one nasty changeset) instead of adding awkward workarounds, but in the meantime, for testing purposes, you can put the code in `npc_spawn_or_marli::Reset` and it should work just as fine.

File: src/server/game/AI/ScriptedAI/ScriptedCreature.cpp (hunk 1 above)
Comment (xvwyh) [reply src/server/game/AI/ScriptedAI/ScriptedCreature.cpp]: I'm afraid the most generic solution you can currently achieve in C++ is not returning anything from the method, but instead make it a template accepting a parameter of arbitrary iterator type, and filling it like `itr++ = guid`. That way users of the method can call it like this:
```cpp
std::vector<ObjectGuid> v;
GetGUIDs(std::back_inserter(v));
std::list<ObjectGuid> l;
GetGUIDs(std::back_inserter(l));
std::set<ObjectGuid> s;
GetGUIDs(std::inserter(s));
```
But I'm not saying you should do it, though. 😄

C#'s IEnumerable, besides being an actual interface (which don't really exist in C++), has an additional benefit of allowing lazy coroutine-like execution: if you write a function that returns IEnumerable and you return data from it via `yield return`, the compiler breaks your function into multiple pieces under the hood (e.g. initialization, loop, conclusion) and only executes each loop iteration when the next element of IEnumerable is requested. That way if you break your loop after the first iteration - only the code until the first `yield return` will end up being executed and nothing else. C++ doesn't have anything like that currently, and while you could certainly code it yourself (by making a custom iterator that calls the appropriate method in `operator++`), you'd have to deal with breaking the function into pieces yourself, which is inconvenient at best (and completely unusable at worst). Coroutine TS [could make it much more pleasant](https://kirit.com/How%20C%2B%2B%20coroutines%20work/Yielding%20Generators) though, when it finally gets included in C++ standard.

**Comment by illfated** [comment]
`npc_spawn_of_marli::Reset` ( or -> of )
(like in the previous comments, just in case you were confused)

**Comment by jackpoz** [comment]
@kuskmen could you work around the "gameobject summoning TempSummon" requirement using GetCreatureListWithEntryInGrid ?

**Comment by kuskmen** [comment]
@jackpoz Thats what I am planning to do, trying to minimize core changes and use whatever is helpful from @xvwyh review, its just I don't have much time right now to sit down and do it all at once.

**Comment by jackpoz** [comment]
no hurry :)

**Comment by xvwyh** [comment]
> @kuskmen could you work around the "gameobject summoning TempSummon" requirement using GetCreatureListWithEntryInGrid, without relying on the hook firing (or not firing) ?

I vote for you to fix #23242 instead. 😄 If you don't fancy the idea of changing summoner from `Unit*` to `WorldObject*` everywhere - heck, just let `IsSummonedBy` be called with `nullptr` instead. After all it's main purpose is only to be called once (unlike `Reset`), and to differentiate summons from static spawns in the world (unlike `Reset` or `InitializeAI` or AI constructor).

Really though, using `GetCreatureListWithEntryInGrid` for that will end up selecting not just newly spawned spiders, but already-spawned-and-not-yet-killed ones too, causing both new and old to aggro on a random target (i.e. existing spiders will randomly switch target), which isn't nice. To mitigate this you'll need to keep track of which spiders have already been handled, and that just adds a lot of unnecessary complexity to a script which could otherwise be dead simple.

**Comment by kuskmen** [comment]
@jackpoz I tried your changes from your branch and I think they worked out the way we planned them to. 

Thanks guys for the reviews (esp. @xvwyh and @jackpoz ) without you I wouldn't understand where I did wrong and unnecessary things, I hope after my last commit you will have to take a look at my changes for the last time.

Given the fact that I merged @jackpoz 's branch into mine I suppose you need to merge his pull request before mine (if future work comes from his branch I will make sure that this branch is up to date with both 3.3.5 and his) or merge everything from here and discard his draft. 

In order to see only my changes they are in the difference of mine and his branch. 
(for the lazy ones: src/server/scripts/EasternKingdoms/ZulGurub/zulgurub.h
src/server/scripts/EasternKingdoms/ZulGurub/instance_zulgurub.cpp
src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp in differences of this pull request)

**Comment by jackpoz** [comment]
https://github.com/jackpoz/TrinityCore/compare/3.3.5-gameobject_summoner...kuskmen:22874-high-priestess-marli-hatching-eggs

This compares this PR to the gameobject summoner one, showing only the changes that you did.

**Comment by kuskmen** [comment]
Ow one more thing, this should be needed as well I think
```sql
INSERT INTO world.spell_script_names VALUES (24083, 'spell_hatch_spiders');
UPDATE world.gameobject_template SET ScriptName = "gob_spider_egg" WHERE entry = 179985;
UPDATE world.gameobject SET ScriptName = "gob_spider_egg" WHERE id = 179985;
```

**Comment by jackpoz** [comment]
could you include that SQL script in the PR, calling it with the usual 9999_99_99_00.sql (or whatever it is we use as placeholder) ? Please don't include "world." in the table name.

**Comment by xvwyh** [comment]
There's no need for `args.AddSpellMod(SPELLVALUE_MAX_TARGETS, 4);`, 4 target limit is already included in the spell. Just `DoCastAOE(SPELL_HATCH_EGGS)` and it should work right out of the box.

`creature->AI()->AttackStart(marli->AI()->SelectTarget(SELECT_TARGET_RANDOM));`
You can somewhat simplify this if you moved it to Mar'li's JustSummoned: `summon->AI()->AttackStart(SelectTarget(SELECT_TARGET_RANDOM));`, but it's not a big deal.

Nice work!

File: sql/updates/world/3.3.5/2019_05_20_00_world.sql [hunk 7]
```diff
@@ -0,0 +1,4 @@
+INSERT INTO spell_script_names VALUES (24083, 'spell_hatch_spiders');
+
+UPDATE gameobject_template SET ScriptName = "gob_spider_egg" WHERE entry = 179985;
+UPDATE gameobject SET ScriptName = "gob_spider_egg" WHERE id = 179985;
```
Comment (Killyana) [inline sql/updates/world/3.3.5/2019_05_20_00_world.sql]: No need to set the scriptname in gameobject table.

**Comment by jackpoz** [comment]
please rebase this PR on latest 335, skipping the merged commits about the GO PR

**Comment by kuskmen** [comment]
@jackpoz tested it, works as intended :shipit:

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Comment by kuskmen** [comment]
@jackpoz @Killyana, by the way, should I cherry-pick this to master as well?

**Comment by Killyana** [comment]
No, cherry-picks must be done in the correct order.
